sensor_control/vision/symbols/raylidar/CameraLidarTransformer.cpp

```diff
@@ -0,0 +1,211 @@
+  visualization_msgs::MarkerArray markers;
+  sensor_msgs::PointCloud2ConstPtr scloud =
+      lidarCache.getElemAfterTime(req.header.stamp);
+  if (!scloud) {
+    res.success = false;
+    res.error = "CLOUD NOT FOUND";
+    return true;
+  }
+  geometry_msgs::TransformStamped transform =
+      tfBuffer.lookupTransform("right_right_cam", "velodyne", req.header.stamp);
+  sensor_msgs::PointCloud2 cloud_transformed;
+  tf2::doTransform(*scloud, cloud_transformed, transform);
+#ifdef DO_ROS_DEBUG
+  debugCloudPub.publish(cloud_transformed);
+  cv::Mat debug_image(camera_info.height, camera_info.width, CV_8UC3,
+                      cv::Scalar(0));
+#endif
+  Eigen::Matrix3d matrixFindPlaneA;
+  matrixFindPlaneA << 0, 0, 0, 0, 0, 0, 0, 0, 0;
+  Eigen::Vector3d matrixFindPlaneB(0, 0, 0);
+
+  cv::circle(debug_image, cv::Point(req.point.x, req.point.y), 3,
+             cv::Scalar(255, 0, 0), 5);
+  //Tracks the closest lidar point to the requested camera point
+  double minDistance = std::numeric_limits<double>::max();
+  for (auto ii = 0, jj = 0; ii < cloud_transformed.width;
+       ++ii, jj += cloud_transformed.point_step) {
+    floatConverter x, y, z, i;
+    for (int kk = 0; kk < 4; ++kk) {
+      x.data[kk] = cloud_transformed.data[jj + kk];
+      y.data[kk] = cloud_transformed.data[jj + 4 + kk];
+      z.data[kk] = cloud_transformed.data[jj + 8 + kk];
+      i.data[kk] = cloud_transformed.data[jj + 16 + kk];
+    }
+    cam_model.fromCameraInfo(camera_info);
+    if (int(z.f) < 0 || int(z.f) > 30) continue;
+    cv::Point2d point = cam_model.project3dToPixel(cv::Point3d(x.f, y.f, z.f));
+
+    if (int(point.x) > 0 && int(point.x) < camera_info.width &&
+        int(point.y) > 0 && int(point.y) < camera_info.height) {
+#ifdef DO_ROS_DEBUG
+      visualization_msgs::Marker marker_point;
+      marker_point.header = req.header;
+      marker_point.header.seq = 0;
+      marker_point.header.frame_id = "right_right_cam";
+      marker_point.id = (int)ii;
+      marker_point.type = visualization_msgs::Marker::CUBE;
+      marker_point.pose.position.x = x.f;
+      marker_point.pose.position.y = y.f;
+      marker_point.pose.position.z = z.f;
+      marker_point.scale.x = 1;
+      marker_point.scale.y = 0.1;
+      marker_point.scale.z = 0.1;
+      marker_point.color.a = 1.0;
+      marker_point.color.r = 0.0;
+      marker_point.color.g = 1.0;
+      marker_point.color.b = 0.0;
+      //marker_point.lifetime = ros::Duration(0.5);
+      markers.markers.push_back(marker_point);
+
+      // Draw
+      if (int(point.x - 20) > 0 && int(point.x + 20) < camera_info.width &&
+          int(point.y - 20) > 0 && int(point.y + 20) < camera_info.height) {
+        cv::circle(debug_image, point, 3,
+                   cv::Scalar(0, 0, 255 * std::min(1.0, z.f / 20.0)), -1);
+      }
+#endif
+      double distance = sqrt(pow(point.x - req.point.x ,2) + pow(point.y - req.point.y,2) ); //Distance (2D) from request point to projected lidar point
```

Comment:
Doesn't really matter here since speed doesn't seem to be a problem but something that will save some time would be not taking the square root and instead making `req.tolerance` in units of distance squared. Since this is getting run on each point, optimizations here effectively get multiplied.